ocaml/xapi-aux/networking_info.ml

```diff
@@ -64,19 +62,14 @@ let get_management_ip_addr ~dbg =
         | "ipv6" ->
             Net.Interface.get_ipv6_addr dbg iface
         | s ->
-            raise
-              (Unexpected_address_type
-                 (Printf.sprintf "Expected 'ipv4' or 'ipv6', got %s" s)
-              )
-      in
-      let addrs =
-        addrs
-        |> List.map (fun (addr, _) -> Unix.string_of_inet_addr addr)
-        |> (* Filter out link-local addresses *)
-        List.filter (fun addr -> String.sub addr 0 4 <> "fe80")
-        |> List.map (fun str ->
-               Option.map (fun bytes -> (str, bytes)) (ip_addr_of_string str)
-           )
+            let msg = Printf.sprintf "Expected 'ipv4' or 'ipv6', got %s" s in
+            L.error "%s: %s" __FUNCTION__ msg ;
+            raise (Unexpected_address_type msg)
       in
-      Option.join (List.nth_opt addrs 0)
+      addrs
+      |> List.map (fun (addr, _) -> Ipaddr_unix.of_inet_addr addr)
+      (* Filter out link-local addresses *)
+      |> List.filter (fun addr -> Ipaddr.scope addr <> Ipaddr.Link)
+      |> List.map (fun ip -> (Ipaddr.to_string ip, ipaddr_to_cstruct ip))
+      |> Fun.flip List.nth_opt 0
```

Comment:
Should we log the complete list before we pick the head as a management address? You could bind `Fun.flip List.nth_opt 0` to `hd_opt` - it returns the head of the list or None.

Reply:
> Should we log the complete list before we pick the head as a management address?

We will probably need to change this selection in the dual-stack case, for the time being it's fine as it is, we can add more debugging when exploring host ipv6.

For the other comment I'm not sure what you mean, there is no `List.hd_opt` in stdlib. Do you mean defining a new binding?

Reply:
The idea was to do `let hd_opt xs = .. in` locally - just for clarity because I had to think for a moment what happens here.